### PR TITLE
API changes and corresponding implementation in the DataPipelineApp for supporting Action plugin.

### PR DESCRIPTION
**Pull request by sagarkapare**
Link to the API design doc - https://wiki.cask.co/display/CE/Custom+actions
Following changes are done as a part of PR:
1. API for the action plugins `Action` and `ActionContext` documented here https://wiki.cask.co/display/CE/Custom+actions
2. Basic implementation of the APIs.
3. DataPipelineApp changes to add the `ETLAction` and ability to configure it.
4. `PipelinePlanner` changes to generate phases and phase connections for the `Action` plugin.
5. Unit test case (will need to add more for the complex pipelines)

### CODE REVIEW

**Comment by sagarkapare** [comment]
@albertshau Please help review the PR. Thanks!

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/DefaultPluginConfigurer.java [hunk 1]
```diff
@@ -43,10 +43,10 @@
  */
 public class DefaultPluginConfigurer extends DefaultDatasetConfigurer implements PluginConfigurer {
 
-  private final Id.Artifact artifactId;
-  private final ArtifactRepository artifactRepository;
-  private final PluginInstantiator pluginInstantiator;
-  private final Map<String, Plugin> plugins;
+  protected final Id.Artifact artifactId;
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/DefaultPluginConfigurer.java]: Why need the change?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/customaction/DefaultCustomActionConfigurer.java [hunk 2]
```diff
@@ -42,7 +46,9 @@
   private String description;
   private Map<String, String> properties;
 
-  private DefaultCustomActionConfigurer(CustomAction customAction) {
+  public DefaultCustomActionConfigurer(CustomAction customAction, Id.Namespace deployNamespace, Id.Artifact artifactId,
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/customaction/DefaultCustomActionConfigurer.java]: should be private.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/DefaultPluginConfigurer.java (hunk 1 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/DefaultPluginConfigurer.java]: Just that it seems like arbitrary that it get exposed in the `DefaultPluginConfigurer`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/workflow/DefaultWorkflowForkConfigurer.java [hunk 3]
```diff
@@ -34,14 +39,16 @@
  * @param <T>
  */
 public class DefaultWorkflowForkConfigurer<T extends WorkflowForkJoiner & WorkflowConditionAdder>
-  implements WorkflowForkConfigurer<T>, WorkflowForkJoiner, WorkflowConditionAdder {
+  extends DefaultPluginConfigurer implements WorkflowForkConfigurer<T>, WorkflowForkJoiner, WorkflowConditionAdder {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/workflow/DefaultWorkflowForkConfigurer.java]: Seems like the extension is for the sake of having `artifactId`, `artifactRepo`... from it. You can just memorize them as fields of this class in the constructor, isn't it? Or the `WorkflowForkConfigurer` actually has the plugin configurer functionalities?

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/DataPipelineApp.java [hunk 4]
```diff
@@ -21,6 +21,7 @@
 import co.cask.cdap.api.dataset.lib.TimePartitionedFileSet;
 import co.cask.cdap.api.schedule.Schedules;
 import co.cask.cdap.etl.api.Transform;
+import co.cask.cdap.etl.api.action.Action;
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/DataPipelineApp.java]: unused?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/DefaultPluginConfigurer.java (hunk 1 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/DefaultPluginConfigurer.java]: Removed the change.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/workflow/DefaultWorkflowForkConfigurer.java (hunk 3 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/workflow/DefaultWorkflowForkConfigurer.java]: No you are right! no need to extend. memoizing now.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/DataPipelineApp.java (hunk 4 above)
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/DataPipelineApp.java]: removed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/customaction/DefaultCustomActionConfigurer.java (hunk 2 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/customaction/DefaultCustomActionConfigurer.java]: This is used by `WorkflowNodeCreator` to create the custom action node. So added it as `public` similar to the `DefaultWorkflowActionConfigurer`.

**Comment by sagarkapare** [comment]
@chtyim @albertshau Addressed comments. Please take a look. Thanks!

**Comment by sagarkapare** [comment]
Build is running here - http://builds.cask.co/browse/CDAP-DUT4382-1

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/customaction/DefaultCustomActionConfigurer.java (hunk 2 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/customaction/DefaultCustomActionConfigurer.java]: I don't see any `new DefaultCustomActionConfigurer` in that class. Do I miss something?

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Arguments.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+  /**
+   * Returns true if arguments has specified option, otherwise false is returned.
+   */
+  boolean hasOption(String name);
+
+  /**
+   * Returns the value for the given option name.
+   */
+  String getOption(String name);
```
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Arguments.java]: can the return value be null ?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/ETLBatchApplication.java [hunk 6]
```diff
@@ -39,6 +39,7 @@
 import org.apache.avro.mapreduce.AvroKeyOutputFormat;
 
 import java.util.HashMap;
+import java.util.HashSet;
```
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/ETLBatchApplication.java]: unused?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/customaction/BasicSettableArguments.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    return options.containsKey(name);
+  }
+
+  @Override
+  public String getOption(String name) {
+    return options.get(name);
+  }
+
+  @Override
+  public Map<String, String> asMap() {
```
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/customaction/BasicSettableArguments.java:54]: `asMap` definition in Arguments says its immutable map, but this `options` map is not immutable.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/customaction/ETLAction.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    .create();
+
+  public ETLAction(BatchPhaseSpec phaseSpec) {
+    this.phaseSpec = phaseSpec;
+  }
+
+  @Override
+  protected void configure() {
+    setName(phaseSpec.getPhaseName());
+    setDescription("CustomAction phase executor. " + phaseSpec.getDescription());
```
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/customaction/ETLAction.java]: what would the description be ? is it same as custom action name

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/ETLBatchApplication.java (hunk 6 above)
Comment (CuriousVini) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/ETLBatchApplication.java]: unused?

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelinePhase.java [hunk 9]
```diff
@@ -68,7 +68,7 @@ private PipelinePhase(Map<String, Set<StageInfo>> stages, Dag dag) {
         stageInfos.addAll(stages.get(pluginType));
       }
     }
-    return Collections.unmodifiableSet(stageInfos == null ? new HashSet<StageInfo>() : stageInfos);
+    return Collections.unmodifiableSet(stageInfos);
```
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelinePhase.java:71]: why this change, stageInfos can't be null ?

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Arguments.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+package co.cask.cdap.etl.api;
+
+import java.util.Map;
+
+/**
+ * Represents arguments passed to the stages in the pipeline.
+ */
+public interface Arguments extends Iterable<Map.Entry<String, String>> {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Arguments.java]: @gsps1 do you know what we're calling the lookup macro? Was it property or argument?  Should just make sure our terminology is consistent.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Arguments.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+  /**
+   * Returns true if arguments has specified option, otherwise false is returned.
+   */
+  boolean hasOption(String name);
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Arguments.java]: I think we can just use 'has' and 'get'. Introducing the term 'Option' may be confusing.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/action/SettableArguments.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+/**
+ * Provides access to the pipeline {@link Arguments} which can be updated.
+ */
+public interface SettableArguments extends Arguments {
+
+  /**
+   * Sets the name and value as specified by the input parameters.
+   */
+  void setOption(String name, String value);
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/action/SettableArguments.java]: think we should just use 'set', as 'Option' may be a confusing term.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/customaction/ETLAction.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+import com.google.gson.GsonBuilder;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Implementation of the {@link CustomAction} used in DataPipeline application.
+ */
+public class ETLAction extends AbstractCustomAction {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/customaction/ETLAction.java]: lets call it PipelineAction or something other than ETL, as we're moving further and further away from purely ETL pipelines.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/PipelinePlanner.java [hunk 14]
```diff
@@ -72,20 +73,60 @@ public PipelinePlan plan(PipelineSpec spec) {
...
       }
+      if (Action.PLUGIN_TYPE.equals(stage.getPlugin().getType())) {
+        // Collect all Action nodes from spec
+        actionNodes.add(stage.getName());
+      }
       specs.put(stage.getName(), stage);
     }
 
+    // Map to hold set of stages to which there is a connection from a action stage.
+    Map<String, Set<String>> outgoingActionConnections = new HashMap<>();
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/PipelinePlanner.java]: may be a little simpler to use a SetMultimap. Then you won't have to do the null checks.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/PipelinePlanner.java [hunk 15]
```diff
@@ -125,10 +166,77 @@ public PipelinePlan plan(PipelineSpec spec) {
...
+          phaseConnections.add(new Connection(connectionFromAction.getKey(), destination));
+        }
+      }
+    }
+
+    // At this point we have build phaseConnections from Action node to another Action node or phaseConnections
+    // from Action node to another subdags. However it is also possible that sudags connects to the action node.
+    // Build those connections here.
+    for (Map.Entry<String, Set<String>> connectionToAction : incomingActionConnections.entrySet()) {
+      // Check if source is one of the source stages in the pipeline
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/PipelinePlanner.java]: seems like it should be // Check if source is one of the sink stages in the pipeline

File: cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/action/MockAction.java [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+@Name("TableWriterAction")
+public class MockAction extends Action {
+  private final Config config;
+  public static final PluginClass PLUGIN_CLASS = getPluginClass();
+
+  /**
+   * Config for the MockAction
+   */
+  public static class Config extends PluginConfig {
+    private String tableName;
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/action/MockAction.java:52]: this mock would be more useful if you can configure the column and value to put. That way you can distinguish which action wrote what.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/customaction/DefaultCustomActionConfigurer.java (hunk 2 above)
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/customaction/DefaultCustomActionConfigurer.java]: WorkflowNodeCreator looks like it uses the static method.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Arguments.java (hunk 10 above)
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Arguments.java]: `lookup` is the  method name with the argument name `property`
https://github.com/caskdata/cdap/blob/develop/cdap-api/src/main/java/co/cask/cdap/api/macro/MacroEvaluator.java#L45

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/action/SettableArguments.java (hunk 12 above)
Comment (gsps1) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/action/SettableArguments.java]: maybe setArgument since the interface is SettableArguments ?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/customaction/DefaultCustomActionConfigurer.java (hunk 2 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/customaction/DefaultCustomActionConfigurer.java]: oh yes. the constructor can be made private, I thought the comment was about the class. my bad.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/ETLBatchApplication.java (hunk 6 above)
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/ETLBatchApplication.java]: Removed.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/customaction/BasicSettableArguments.java (hunk 7 above)
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/customaction/BasicSettableArguments.java:54]: yeah. I removed the `Arguments` interface since it is unused for now and having `SettableArguments` interface which returns the mutable map.

File: cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/action/MockAction.java (hunk 16 above)
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/action/MockAction.java:52]: Added the configuration for rowKey, columnKey and value.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/PipelinePlanner.java (hunk 15 above)
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/PipelinePlanner.java]: good catch. fixed.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/PipelinePlanner.java (hunk 14 above)
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/PipelinePlanner.java]: using SetMultiMap

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/customaction/ETLAction.java (hunk 13 above)
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/customaction/ETLAction.java]: Calling PipelineAction.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/action/SettableArguments.java (hunk 12 above)
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/action/SettableArguments.java]: Removed `Options` from the API.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Arguments.java (hunk 11 above)
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Arguments.java]: fixed.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Arguments.java (hunk 10 above)
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Arguments.java]: Let me know if should be renamed to something like `SettablePipelineProperties`. However `SettableArguments` makes more sense I think in this case.

**Comment by sagarkapare** [comment]
@albertshau @chtyim @gsps1 Please help review the PR. Thanks!

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/action/ActionContext.java [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+   * started by the scheduler. Otherwise it would be the current time when the action runs.
+   *
+   * @return Time in milliseconds since epoch time (00:00:00 January 1, 1970 UTC).
+   */
+  long getLogicalStartTime();
+
+  /**
+   * Return the arguments which can be updated.
+   */
+  SettableArguments getArguments();
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/action/ActionContext.java:39]: Do we call it arguments now? I thought we said it's properties. Don't have strong opinion on this.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/action/SettableArguments.java [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+ */
+package co.cask.cdap.etl.api.action;
+
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Provides access to the pipeline arguments which can be updated.
+ */
+public interface SettableArguments extends Iterable<Map.Entry<String, String>> {
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/action/SettableArguments.java:24]: It's a bit weird that there is a `SettableArguments` but not corresponding `Arguments`. Also, maybe better call it `MutableArguments` as settable is not really a word.

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/customaction/PipelineAction.java [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+    StageInfo stageInfo = phase.iterator().next();
+    Action action = context.newPluginInstance(stageInfo.getName());
+    ActionContext actionContext = new BasicActionContext(context);
+    action.run(actionContext);
+    WorkflowToken token = context.getWorkflowToken();
+    if (token == null) {
+      throw new IllegalStateException("WorkflowToken cannot be null when action is executed through Workflow.");
+    }
+    for (Map.Entry<String, String> entry : actionContext.getArguments()) {
+      context.getWorkflowToken().put(entry.getKey(), entry.getValue());
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/customaction/PipelineAction.java]: nit: can use token instead of context.getWorkflowToken()

**Comment by albertshau** [comment]
minor comments, otherwise lgtm

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/action/ActionContext.java (hunk 17 above)
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/action/ActionContext.java:39]: Using arguments because UI calls it as an arguments.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/action/SettableArguments.java (hunk 18 above)
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/action/SettableArguments.java:24]: Currently only `SettableArguments` is used (by `Action` plugins). `Arguments` can be added later if required. Since this interface exposes `set` method calling the class as `SettableArguments` similar to the `SettableFuture`. However no strong opinion here, please let me know if name should be changed to `MutableArgument`.

**Comment by sagarkapare** [comment]
Build is green - http://builds.cask.co/browse/CDAP-DUT4382-4. Squashing and merging. @chtyim Please let me know if the name of the `SettableArguments` should be changed to `MutableArguments`. Would do it as a separate PR. Thanks!
